File: simulate.py
"""Monte Carlo tournament simulator using ELO ratings.

Goal model: independent Poisson per team.
    xg_a = max(0.25, 1.35 + (elo_a - elo_b) * 0.005)
    xg_b = max(0.25, 1.35 - (elo_a - elo_b) * 0.005)

So a 200-ELO gap shifts expected GD by ~1 goal. Knockout draws go to extra
time (1/3 of regulation goal expectation) and then a near-coin-flip shootout
slightly weighted by ELO.
"""
from __future__ import annotations

import argparse
import math
import random
from collections import Counter, defaultdict
from dataclasses import dataclass
from pathlib import Path

import yaml

BASE_XG = 1.35
ELO_TO_XG = 0.0025         # tuned to match per-match W/D/L on 51K historical
                            # matches (bucket calibration shows 0.003 was ~3-4pp
                            # over-confident at the favorite at moderate diffs)
MIN_XG = 0.25
ET_FRACTION = 1 / 3        # 30 of 90 minutes
PK_ELO_DAMP = 1200         # penalty shootouts: 1200-ELO scale (very dampened)
HOST_ADVANTAGE_DEFAULT = 50  # ELO bonus for host country teams


def effective_elos(config: dict) -> dict[str, int]:
    """Return {code: elo + host_bonus_if_host}."""
    hosts = set(config.get("hosts") or [])
    boost = config.get("host_advantage")
    if boost is None:
        boost = HOST_ADVANTAGE_DEFAULT
    return {
        code: info["elo"] + (boost if code in hosts else 0)
        for code, info in config["teams"].items()
    }


GLOBAL_CRITERIA = {"points", "gd", "gf"}
H2H_CRITERIA = {"h2h_points", "h2h_gd", "h2h_gf"}


@dataclass
class Match:
    home: str
    away: str
    home_goals: int
    away_goals: int


def poisson(mean: float) -> int:
    """Knuth's algorithm; fine for the small means we use."""
    L = math.exp(-mean)
    k, p = 0, 1.0
    while True:
        k += 1
        p *= random.random()
        if p <= L:
            return k - 1


def regulation_goals(elo_a: int, elo_b: int) -> tuple[int, int]:
    xg_a = max(MIN_XG, BASE_XG + (elo_a - elo_b) * ELO_TO_XG)
    xg_b = max(MIN_XG, BASE_XG - (elo_a - elo_b) * ELO_TO_XG)
    return poisson(xg_a), poisson(xg_b)


def knockout_winner(name_a: str, elo_a: int, name_b: str, elo_b: int) -> tuple[str, str]:
    """Returns (winner, loser). Tied regulations resolve by coin flip — stands
    in for extra time + penalty shootouts (empirically close to 50/50) and
    absorbs day-of-form variance the per-match model can't see."""
    ga, gb = regulation_goals(elo_a, elo_b)
    if ga != gb:
        return (name_a, name_b) if ga > gb else (name_b, name_a)
    return (name_a, name_b) if random.random() < 0.5 else (name_b, name_a)


def team_record(team: str, matches: list[Match]) -> dict[str, int]:
    pts = gf = ga = 0
    for m in matches:
        if m.home == team:
            gf += m.home_goals; ga += m.away_goals
            if m.home_goals > m.away_goals: pts += 3
            elif m.home_goals == m.away_goals: pts += 1
        elif m.away == team:
            gf += m.away_goals; ga += m.home_goals
            if m.away_goals > m.home_goals: pts += 3
            elif m.away_goals == m.home_goals: pts += 1
    return {"points": pts, "gd": gf - ga, "gf": gf}


def criterion_value(crit: str, team: str, matches: list[Match]) -> float:
    if crit == "random":
        return random.random()
    key = crit.replace("h2h_", "")
    return team_record(team, matches)[key]


def rank_group(teams: list[str], matches: list[Match], criteria: list[str]) -> list[str]:
    """Sort teams best-first by criteria. h2h_* criteria restrict matches to
    those among the currently-tied subgroup."""
    if len(teams) <= 1 or not criteria:
        return list(teams)
    head, *rest = criteria
    scope = matches
    if head in H2H_CRITERIA:
        team_set = set(teams)
        scope = [m for m in matches if m.home in team_set and m.away in team_set]
    keys = {t: criterion_value(head, t, scope) for t in teams}
    buckets: dict = defaultdict(list)
    for t in teams:
        buckets[keys[t]].append(t)
    out: list[str] = []
    for k in sorted(buckets.keys(), reverse=True):
        bucket = buckets[k]
        out.extend(bucket if len(bucket) == 1 else rank_group(bucket, matches, rest))
    return out


def resolve(ref: str, slots: dict[str, str], teams: dict) -> str:
    if ref in teams:
        return ref
    if ref in slots:
        return slots[ref]
    raise KeyError(f"unresolved slot reference: {ref}")


def teams_in_stage(stage: dict, slots: dict, teams: dict) -> set[str]:
    t = set()
    if stage["type"] == "groups":
        for grp in stage["groups"].values():
            t.update(resolve(x, slots, teams) for x in grp)
    elif stage["type"] == "bracket":
        for m in stage["matches"].values():
            t.add(resolve(m["home"], slots, teams))
            t.add(resolve(m["away"], slots, teams))
    elif stage["type"] == "match":
        t.add(resolve(stage["home"], slots, teams))
        t.add(resolve(stage["away"], slots, teams))
    # thirds_pool stages don't introduce new teams; they redirect existing ones.
    return t


def thirds_assign(stage: dict, slots: dict, team_stats: dict, source_stage: dict) -> None:
    """Rank source_stage's Nth-place teams across groups (points → GD → GF →
    random) and fill stage.id slots either by the per-combo assignment table
    or, if no entry matches, by raw rank (slot names 1..take)."""
    candidates = []
    for group_name in source_stage["groups"]:
        team = slots.get(f"{stage['source']}.{group_name}.{stage['rank']}")
        if not team:
            continue
        stats = team_stats.get(team, {"points": 0, "gd": 0, "gf": 0})
        candidates.append({"team": team, "group": group_name, **stats})
    candidates.sort(
        key=lambda c: (-c["points"], -c["gd"], -c["gf"], random.random())
    )
    top = candidates[: stage["take"]]
    combo = "".join(sorted(c["group"] for c in top))
    assignment = (stage.get("assignments") or {}).get(combo)
    if assignment:
        for slot_name, source_group in assignment.items():
            c = next((x for x in candidates if x["group"] == source_group), None)
            if c:
                slots[f"{stage['id']}.{slot_name}"] = c["team"]
    else:
        for i, c in enumerate(top, start=1):
            slots[f"{stage['id']}.{i}"] = c["team"]


def simulate_once(config: dict) -> tuple[dict[str, str], dict[str, set[str]]]:
    teams = config["teams"]
    criteria = config.get("tiebreakers",
                          ["points", "gd", "gf", "h2h_points", "h2h_gd", "random"])
    slots: dict[str, str] = {}
    reached: dict[str, set[str]] = {}
    team_stats: dict[str, dict] = {}
    elo = effective_elos(config)

    for stage in config["stages"]:
        sid = stage["id"]
        reached[sid] = teams_in_stage(stage, slots, teams)

        if stage["type"] == "groups":
            pairings = stage.get("pairings") or {}
            for group_name, codes in stage["groups"].items():
                actual = [resolve(c, slots, teams) for c in codes]
                if group_name in pairings:
                    pairs = [
                        (resolve(a, slots, teams), resolve(b, slots, teams))
                        for a, b in pairings[group_name]
                    ]
                else:
                    pairs = [
                        (actual[i], actual[j])
                        for i in range(len(actual))
                        for j in range(i + 1, len(actual))
                    ]
                matches: list[Match] = []
                for a, b in pairs:
                    ga, gb = regulation_goals(elo[a], elo[b])
                    matches.append(Match(a, b, ga, gb))
                for code in actual:
                    rec = team_record(code, matches)
                    team_stats[code] = {**rec, "group": group_name}
                ranked = rank_group(actual, matches, criteria)
                for pos, code in enumerate(ranked, start=1):
                    slots[f"{sid}.{group_name}.{pos}"] = code

        elif stage["type"] == "thirds_pool":
            source = next((s for s in config["stages"] if s["id"] == stage["source"]), None)
            if source is None:
                raise KeyError(f"thirds_pool: source stage {stage['source']!r} not found")
            thirds_assign(stage, slots, team_stats, source)

        elif stage["type"] == "bracket":
            for m_id, m_def in stage["matches"].items():
                a = resolve(m_def["home"], slots, teams)
                b = resolve(m_def["away"], slots, teams)
                w, l = knockout_winner(a, elo[a], b, elo[b])
                slots[f"{sid}.{m_id}.W"] = w
                slots[f"{sid}.{m_id}.L"] = l

        elif stage["type"] == "match":
            a = resolve(stage["home"], slots, teams)
            b = resolve(stage["away"], slots, teams)
            w, l = knockout_winner(a, elo[a], b, elo[b])
            slots[f"{sid}.W"] = w
            slots[f"{sid}.L"] = l

    return slots, reached


def simulate(config: dict, n: int) -> dict[str, Counter]:
    """Returns {team_code: Counter({stage_id: count, 'champion': count})}."""
    stage_ids = [s["id"] for s in config["stages"]]
    champion_slot = config.get("champion_slot") or f"{stage_ids[-1]}.W"
    stats: dict[str, Counter] = defaultdict(Counter)

    for _ in range(n):
        slots, reached = simulate_once(config)
        for sid, teams_set in reached.items():
            for t in teams_set:
                stats[t][sid] += 1
        if champion_slot in slots:
            stats[slots[champion_slot]]["champion"] += 1
    return stats


def main() -> None:
    p = argparse.ArgumentParser()
    p.add_argument("yaml_file", type=Path)
    p.add_argument("--n", type=int, default=10000)
    p.add_argument("--seed", type=int, default=None)
    args = p.parse_args()

    if args.seed is not None:
        random.seed(args.seed)

    config = yaml.safe_load(args.yaml_file.read_text())
    stage_ids = [s["id"] for s in config["stages"]]
    stats = simulate(config, args.n)

    print(f"\n{config['name']}  —  {args.n} simulations\n")
    # Skip the group stage (everyone reaches) and any third_place stage (noisy).
    cols = [s for s in stage_ids[1:] if s != "third_place"] + ["champion"]
    width = max(len(c) for c in cols) + 2
    header = f"  {'Team':<28}" + "".join(f"{c:>{width}}" for c in cols)
    print(header)
    print("  " + "-" * (len(header) - 2))

    by_champion = sorted(
        config["teams"].items(),
        key=lambda kv: -stats[kv[0]]["champion"],
    )
    for code, info in by_champion:
        cells = "".join(f"{stats[code][c] / args.n * 100:>{width - 1}.1f}%" for c in cols)
        print(f"  {info['name']:<28}{cells}")


if __name__ == "__main__":
    main()

File: static/simulate.js
// Monte Carlo tournament simulator. Port of simulate.py.
// Works in browser (exposes window.Football) and node (module.exports).
(function (root) {
  const BASE_XG = 1.35;
  const ELO_TO_XG = 0.0025;   // per-match calibration on 51K historical matches
  const MIN_XG = 0.25;
  const ET_FRACTION = 1 / 3;
  const PK_ELO_DAMP = 1200;
  const HOST_ADVANTAGE_DEFAULT = 50;   // ELO bonus for host country teams

  const DEFAULT_MODEL = Object.freeze({
    eloStrength: 1,
    hostAdvantage: HOST_ADVANTAGE_DEFAULT,
    laterRoundEffect: 1,
    recency: 0,
  });

  function modelSettings(settings) {
    return { ...DEFAULT_MODEL, ...(settings || {}) };
  }

  function effectiveElos(config, settings) {
    // Returns { code: effective_elo }. Adds the host bonus when the team
    // is listed in config.hosts.
    const model = modelSettings(settings);
    const hosts = new Set(config.hosts || []);
    const boost = config.host_advantage != null
      ? config.host_advantage : model.hostAdvantage;
    const out = {};
    for (const [code, info] of Object.entries(config.teams)) {
      const trend = info.elo_9m == null ? 0 : info.elo - info.elo_9m;
      out[code] = info.elo + trend * model.recency + (hosts.has(code) ? boost : 0);
    }
    return out;
  }

  const H2H = new Set(['h2h_points', 'h2h_gd', 'h2h_gf']);
  const DEFAULT_CRITERIA = ['points', 'gd', 'gf', 'h2h_points', 'h2h_gd', 'random'];

  function poisson(mean) {
    const L = Math.exp(-mean);
    let k = 0, p = 1.0;
    while (true) {
      k += 1;
      p *= Math.random();
      if (p <= L) return k - 1;
    }
  }

  function stageSlopeFactor(stageId, settings) {
    // Late-stage fatigue / pressure: dampen ELO weight so favorites have
    // less advantage as the tournament wears on. laterRoundEffect=0 removes
    // the effect; 1 is the calibrated model; 2 doubles it.
    const effect = modelSettings(settings).laterRoundEffect;
    if (stageId === 'final') return Math.max(0, 1 - 0.5 * effect);
    if (stageId === 'sf' || stageId === 'third_place') return Math.max(0, 1 - 0.3 * effect);
    return 1.0;
  }

  function expectedGoals(eloA, eloB, slopeFactor = 1.0, settings) {
    const eff = ELO_TO_XG * modelSettings(settings).eloStrength * slopeFactor;
    const xgA = Math.max(MIN_XG, BASE_XG + (eloA - eloB) * eff);
    const xgB = Math.max(MIN_XG, BASE_XG - (eloA - eloB) * eff);
    return [xgA, xgB];
  }

  function regulationGoals(eloA, eloB, slopeFactor = 1.0, settings) {
    const [xgA, xgB] = expectedGoals(eloA, eloB, slopeFactor, settings);
    return [poisson(xgA), poisson(xgB)];
  }

  function knockoutWinner(nameA, eloA, nameB, eloB, slopeFactor = 1.0, settings) {
    const [ga, gb] = regulationGoals(eloA, eloB, slopeFactor, settings);
    if (ga !== gb) return ga > gb ? [nameA, nameB] : [nameB, nameA];
    // Tied after regulation: coin flip. This stands in for extra time and
    // penalty shootouts; empirically those outcomes are close to 50/50
    // regardless of ELO, and the coin flip also absorbs "form of the day"
    // variance our model otherwise can't see.
    return Math.random() < 0.5 ? [nameA, nameB] : [nameB, nameA];
  }

  function outcomeProbabilitiesFromXg(xgA, xgB) {
    const maxGoals = 20;
    const distributions = [xgA, xgB].map(mean => {
      const out = [Math.exp(-mean)];
      for (let k = 1; k <= maxGoals; k++) out.push(out[k - 1] * mean / k);
      return out;
    });
    let win = 0, draw = 0, loss = 0;
    for (let a = 0; a <= maxGoals; a++) {
      for (let b = 0; b <= maxGoals; b++) {
        const p = distributions[0][a] * distributions[1][b];
        if (a > b) win += p;
        else if (a === b) draw += p;
        else loss += p;
      }
    }
    const total = win + draw + loss;
    return { win: win / total, draw: draw / total, loss: loss / total };
  }

  function outcomeProbabilities(eloDiff, settings, slopeFactor = 1.0) {
    const [xgA, xgB] = expectedGoals(eloDiff, 0, slopeFactor, settings);
    return outcomeProbabilitiesFromXg(xgA, xgB);
  }

  // --- Score prediction -----------------------------------------------------
  // simulate() samples to estimate how often teams reach each round. The
  // functions below answer a different question: the single most-likely
  // *scoreline* of one match — what you need to fill in a score-prediction
  // pool. Mirror of predict.py; keep the two in sync.

  const BLOWOUT_GAP_DEFAULT = 440;    // raw ELO gap that calls an outright 4-0
  const MARKET_WEIGHT_DEFAULT = 0.8;  // weight on market xG when blending

  function marketXg(ah, ou) {
    // Convert bookmaker lines to expected goals for [home, away]. The
    // Asian-handicap line is the negative of the home team's supremacy and the
    // over/under line is the expected total: xgHome - xgAway = -ah,
    // xgHome + xgAway = ou.
    const supremacy = -ah;
    return [
      Math.max(MIN_XG, (ou + supremacy) / 2),
      Math.max(MIN_XG, (ou - supremacy) / 2),
    ];
  }

  function poissonMode(mean) {
    // Mode of a Poisson(mean): floor(mean), floored at our minimum xG.
    return Math.floor(Math.max(MIN_XG, mean));
  }

  function scorelineProbabilities(xgA, xgB, top = null, maxGoals = 10) {
    // Scorelines ranked by probability (independent Poissons). With top set,
    // return only that many of the most likely ones.
    const dists = [xgA, xgB].map(mean => {
      const out = [Math.exp(-mean)];
      for (let k = 1; k <= maxGoals; k++) out.push(out[k - 1] * mean / k);
      return out;
    });
    const grid = [];
    for (let a = 0; a <= maxGoals; a++)
      for (let b = 0; b <= maxGoals; b++)
        grid.push([[a, b], dists[0][a] * dists[1][b]]);
    grid.sort((x, y) => y[1] - x[1]);
    return top ? grid.slice(0, top) : grid;
  }

  function outcomeAwareScore(xgA, xgB) {
    // Most likely scoreline consistent with the most likely result. The plain
    // modal score over-predicts draws; conditioning on the win/draw/loss
    // outcome first scores better in pool backtests.
    const { win, draw, loss } = outcomeProbabilitiesFromXg(xgA, xgB);
    const best = (win >= draw && win >= loss) ? 'w' : (draw >= loss ? 'd' : 'l');
    const ok = ([a, b]) => best === 'w' ? a > b : best === 'd' ? a === b : a < b;
    for (const [s] of scorelineProbabilities(xgA, xgB)) if (ok(s)) return s;
    return [1, 1];
  }

  function predictedScore(eloA, eloB, options = {}) {
    // Most-likely scoreline for A vs B. eloA/eloB are effective ELOs (host
    // bonus included) and drive the xG model. Options:
    //   market       {ah, ou}  blend market xG with model xG before the mode
    //   marketWeight  0..1      weight on market xG (default 0.8)
    //   blowout       number    raw ELO gap for an outright 4-0 (null disables)
    //   rawDiff       number    no-host gap gating the blowout (default eloA-eloB)
    const {
      market = null,
      marketWeight = MARKET_WEIGHT_DEFAULT,
      blowout = BLOWOUT_GAP_DEFAULT,
      rawDiff = eloA - eloB,
      outcomeAware = false,
      top = null,
      slopeFactor = 1.0,
      settings,
    } = options;
    let [xgA, xgB] = expectedGoals(eloA, eloB, slopeFactor, settings);
    let blended = false;
    if (market) {
      const [mA, mB] = marketXg(market.ah, market.ou);
      xgA = marketWeight * mA + (1 - marketWeight) * xgA;
      xgB = marketWeight * mB + (1 - marketWeight) * xgB;
      blended = true;
    }
    let [a, b] = outcomeAware ? outcomeAwareScore(xgA, xgB)
                              : [poissonMode(xgA), poissonMode(xgB)];
    let isBlowout = false;
    if (blowout != null) {
      if (rawDiff >= blowout) { a = 4; b = 0; isBlowout = true; }
      else if (rawDiff <= -blowout) { a = 0; b = 4; isBlowout = true; }
    }
    const result = {
      score: [a, b],
      xg: [xgA, xgB],
      probs: outcomeProbabilitiesFromXg(xgA, xgB),
      market: blended,
      blowout: isBlowout,
    };
    if (top) {
      result.top = scorelineProbabilities(xgA, xgB, top)
        .map(([s, p]) => [s, Math.round(p * 1e4) / 1e4]);
    }
    return result;
  }

  function teamRecord(team, matches) {
    let pts = 0, gf = 0, ga = 0;
    for (const m of matches) {
      if (m.home === team) {
        gf += m.homeGoals; ga += m.awayGoals;
        if (m.homeGoals > m.awayGoals) pts += 3;
        else if (m.homeGoals === m.awayGoals) pts += 1;
      } else if (m.away === team) {
        gf += m.awayGoals; ga += m.homeGoals;
        if (m.awayGoals > m.homeGoals) pts += 3;
        else if (m.awayGoals === m.homeGoals) pts += 1;
      }
    }
    return { points: pts, gd: gf - ga, gf };
  }

  function criterionValue(crit, team, matches) {
    if (crit === 'random') return Math.random();
    const key = crit.replace('h2h_', '');
    return teamRecord(team, matches)[key];
  }

  function rankGroup(teams, matches, criteria) {
    if (teams.length <= 1 || criteria.length === 0) return [...teams];
    const [head, ...rest] = criteria;
    let scope = matches;
    if (H2H.has(head)) {
      const ts = new Set(teams);
      scope = matches.filter(m => ts.has(m.home) && ts.has(m.away));
    }
    const buckets = new Map();
    for (const t of teams) {
      const k = criterionValue(head, t, scope);
      if (!buckets.has(k)) buckets.set(k, []);
      buckets.get(k).push(t);
    }
    const out = [];
    for (const k of [...buckets.keys()].sort((a, b) => b - a)) {
      const bucket = buckets.get(k);
      if (bucket.length === 1) out.push(bucket[0]);
      else out.push(...rankGroup(bucket, matches, rest));
    }
    return out;
  }

  function resolve(ref, slots, teams) {
    if (ref in teams) return ref;
    if (slots.has(ref)) return slots.get(ref);
    throw new Error(`unresolved slot: ${ref}`);
  }

  function teamsInStage(stage, slots, teams) {
    const out = new Set();
    if (stage.type === 'groups') {
      for (const grp of Object.values(stage.groups)) {
        for (const c of grp) out.add(resolve(c, slots, teams));
      }
    } else if (stage.type === 'bracket') {
      for (const m of Object.values(stage.matches)) {
        out.add(resolve(m.home, slots, teams));
        out.add(resolve(m.away, slots, teams));
      }
    } else if (stage.type === 'match') {
      out.add(resolve(stage.home, slots, teams));
      out.add(resolve(stage.away, slots, teams));
    }
    return out;
  }

  function thirdsAssign(stage, slots, teamStats) {
    // Collect Nth-place teams from each source group, with their stats.
    const sourceId = stage.source;
    const sourceStage = stage._sourceStage;
    const candidates = [];
    for (const groupName of Object.keys(sourceStage.groups)) {
      const team = slots.get(`${sourceId}.${groupName}.${stage.rank}`);
      if (!team) continue;
      const stats = teamStats.get(team) || { points: 0, gd: 0, gf: 0 };
      candidates.push({ team, group: groupName, ...stats });
    }
    candidates.sort((a, b) =>
      b.points - a.points || b.gd - a.gd || b.gf - a.gf || Math.random() - 0.5);
    const top = candidates.slice(0, stage.take);
    const combo = top.map(c => c.group).sort().join('');
    const assignment = stage.assignments && stage.assignments[combo];
    if (assignment) {
      for (const [slotName, sourceGroup] of Object.entries(assignment)) {
        const c = candidates.find(c => c.group === sourceGroup);
        if (c) slots.set(`${stage.id}.${slotName}`, c.team);
      }
    } else {
      // Fallback: ordered assignment by rank (best third → thirds.1, etc.)
      top.forEach((c, i) => slots.set(`${stage.id}.${i + 1}`, c.team));
    }
  }

  function applySlotOverrides(stage, slots, config) {
    // Pin a slot to a specific team and swap whatever was there into the
    // other team's old position. Used for tournaments where reality
    // can't be derived from match data alone — e.g. 1954's group 2 had
    // DE-TR tied on points with no FIFA-recognised statistical tiebreaker,
    // and the playoff itself decided it.
    if (!config.slot_overrides) return;
    for (const [slot, team] of Object.entries(config.slot_overrides)) {
      if (!slot.startsWith(stage.id + '.')) continue;
      const current = slots.get(slot);
      if (current === team || current === undefined) continue;
      const groupPrefix = slot.split('.').slice(0, -1).join('.') + '.';
      let otherSlot = null;
      for (const [k, v] of slots) {
        if (v === team && k.startsWith(groupPrefix)) { otherSlot = k; break; }
      }
      if (otherSlot) {
        slots.set(slot, team);
        slots.set(otherSlot, current);
      }
    }
  }

  function simulateOnce(config, overrides, settings) {
    const { teams } = config;
    const criteria = config.tiebreakers || DEFAULT_CRITERIA;
    const elo = effectiveElos(config, settings);   // ELO + host bonus per team
    const slots = new Map();
    const reached = new Map();
    const teamStats = new Map();   // team code → {points, gd, gf}

    for (const stage of config.stages) {
      reached.set(stage.id, teamsInStage(stage, slots, teams));

      if (stage.type === 'groups') {
        for (const [name, codes] of Object.entries(stage.groups)) {
          const actual = codes.map(c => resolve(c, slots, teams));
          const matches = [];
          // Use explicit pairings if declared (seeded-group formats like
          // WC 1954); otherwise default to full round-robin.
          let pairs;
          if (stage.pairings && stage.pairings[name]) {
            pairs = stage.pairings[name].map(([a, b]) =>
              [resolve(a, slots, teams), resolve(b, slots, teams)]);
          } else {
            pairs = [];
            for (let i = 0; i < actual.length; i++) {
              for (let j = i + 1; j < actual.length; j++) {
                pairs.push([actual[i], actual[j]]);
              }
            }
          }
          for (const [a, b] of pairs) {
            const fixed = overrides && overrides[`${stage.id}.${a}.${b}`];
            const [ga, gb] = fixed || regulationGoals(elo[a], elo[b], 1, settings);
            matches.push({ home: a, away: b, homeGoals: ga, awayGoals: gb });
          }
          // Record cross-group-comparable stats before ranking.
          for (const code of actual) {
            teamStats.set(code, { ...teamRecord(code, matches), group: name });
          }
          const ranked = rankGroup(actual, matches, criteria);
          ranked.forEach((code, i) => slots.set(`${stage.id}.${name}.${i + 1}`, code));
        }
      } else if (stage.type === 'thirds_pool') {
        // Cache resolved source stage on first encounter.
        if (!stage._sourceStage) {
          stage._sourceStage = config.stages.find(s => s.id === stage.source);
          if (!stage._sourceStage) {
            throw new Error(`thirds_pool: source stage "${stage.source}" not found`);
          }
        }
        thirdsAssign(stage, slots, teamStats);
      } else if (stage.type === 'bracket') {
        for (const [mid, mdef] of Object.entries(stage.matches)) {
          const a = resolve(mdef.home, slots, teams);
          const b = resolve(mdef.away, slots, teams);
          const ovr = overrides && overrides[`${stage.id}.${mid}`];
          let w, l;
          if (ovr && (ovr.winner === a || ovr.winner === b)) {
            w = ovr.winner;
            l = w === a ? b : a;
          } else {
            [w, l] = knockoutWinner(
              a, elo[a], b, elo[b], stageSlopeFactor(stage.id, settings), settings);
          }
          slots.set(`${stage.id}.${mid}.W`, w);
          slots.set(`${stage.id}.${mid}.L`, l);
        }
      } else if (stage.type === 'match') {
        const a = resolve(stage.home, slots, teams);
        const b = resolve(stage.away, slots, teams);
        const ovr = overrides && overrides[stage.id];
        let w, l;
        if (ovr && (ovr.winner === a || ovr.winner === b)) {
          w = ovr.winner;
          l = w === a ? b : a;
        } else {
          [w, l] = knockoutWinner(
            a, elo[a], b, elo[b], stageSlopeFactor(stage.id, settings), settings);
        }
        slots.set(`${stage.id}.W`, w);
        slots.set(`${stage.id}.L`, l);
      }
      applySlotOverrides(stage, slots, config);
    }
    return { slots, reached };
  }

  function simulateInto(config, n, overrides, settings, stats = new Map()) {
    const stageIds = config.stages.map(s => s.id);
    const championSlot = config.champion_slot || `${stageIds[stageIds.length - 1]}.W`;
    const inc = (team, key) => {
      if (!stats.has(team)) stats.set(team, new Map());
      const c = stats.get(team);
      c.set(key, (c.get(key) || 0) + 1);
    };
    for (let i = 0; i < n; i++) {
      const { slots, reached } = simulateOnce(config, overrides, settings);
      for (const [sid, ts] of reached) for (const t of ts) inc(t, sid);
      const champ = slots.get(championSlot);
      if (champ) inc(champ, 'champion');
    }
    return stats;
  }

  function simulate(config, n, overrides, settings) {
    return simulateInto(config, n, overrides, settings);
  }

  // Given the current overrides, lock slots that resolve deterministically.
  // Used by the UI to know which knockout match participants are settled.
  // Excludes the `random` tiebreaker — true ties leave teams in YAML order.
  function resolveDeterministic(config, overrides) {
    const { teams } = config;
    const criteria = (config.tiebreakers || DEFAULT_CRITERIA).filter(c => c !== 'random');
    const slots = new Map();
    const teamStats = new Map();
    const tryResolve = (ref) => {
      if (ref in teams) return ref;
      if (slots.has(ref)) return slots.get(ref);
      return null;
    };
    for (const stage of config.stages) {
      if (stage.type === 'groups') {
        for (const [name, codes] of Object.entries(stage.groups)) {
          const actual = codes.map(tryResolve);
          if (actual.some(x => x === null)) continue;
          const matches = [];
          let allFixed = true;
          for (let i = 0; i < actual.length && allFixed; i++) {
            for (let j = i + 1; j < actual.length; j++) {
              const a = actual[i], b = actual[j];
              const fixed = overrides && overrides[`${stage.id}.${a}.${b}`];
              if (!fixed) { allFixed = false; break; }
              matches.push({ home: a, away: b, homeGoals: fixed[0], awayGoals: fixed[1] });
            }
          }
          if (!allFixed) continue;
          for (const code of actual) {
            teamStats.set(code, { ...teamRecord(code, matches), group: name });
          }
          const ranked = rankGroup(actual, matches, criteria);
          ranked.forEach((c, i) => slots.set(`${stage.id}.${name}.${i + 1}`, c));
        }
      } else if (stage.type === 'thirds_pool') {
        if (!stage._sourceStage) {
          stage._sourceStage = config.stages.find(s => s.id === stage.source);
          if (!stage._sourceStage) continue;
        }
        // Only resolve if every source group's Nth slot is set.
        const sourceGroups = Object.keys(stage._sourceStage.groups);
        const ready = sourceGroups.every(g => slots.has(`${stage.source}.${g}.${stage.rank}`));
        if (!ready) continue;
        thirdsAssign(stage, slots, teamStats);
      } else if (stage.type === 'bracket') {
        for (const [mid, mdef] of Object.entries(stage.matches)) {
          const a = tryResolve(mdef.home);
          const b = tryResolve(mdef.away);
          if (a === null || b === null) continue;
          const ovr = overrides && overrides[`${stage.id}.${mid}`];
          if (!ovr || (ovr.winner !== a && ovr.winner !== b)) continue;
          slots.set(`${stage.id}.${mid}.W`, ovr.winner);
          slots.set(`${stage.id}.${mid}.L`, ovr.winner === a ? b : a);
        }
      } else if (stage.type === 'match') {
        const a = tryResolve(stage.home);
        const b = tryResolve(stage.away);
        if (a === null || b === null) continue;
        const ovr = overrides && overrides[stage.id];
        if (!ovr || (ovr.winner !== a && ovr.winner !== b)) continue;
        slots.set(`${stage.id}.W`, ovr.winner);
        slots.set(`${stage.id}.L`, ovr.winner === a ? b : a);
      }
    }
    return slots;
  }

  const api = {
    DEFAULT_MODEL, effectiveElos, simulate, simulateInto, simulateOnce, regulationGoals, knockoutWinner,
    outcomeProbabilities, rankGroup, resolveDeterministic, stageSlopeFactor,
    marketXg, poissonMode, predictedScore, scorelineProbabilities, outcomeAwareScore,
  };
  if (typeof module !== 'undefined' && module.exports) module.exports = api;
  else root.Football = api;
})(typeof globalThis !== 'undefined' ? globalThis : this);
